### DISCUSSION TITLE
Namespaces

### DISCUSSION THREAD

**Discussion by cristianbuse** [Theadd/ds3x#26] [General]
Hi @Theadd ,

I noticed that you've converted standard modules into classes, like ```JSON``` and ```MemoryLib``` and turned the ```Attribute VB_PredeclaredId``` to ```True```.

My assumption is that you wanted to enforce the full name as in ```JSON.Parse``` hence avoiding to pollute the global namespace.

I used to do this as well, many years ago. I decided against it for the following reasons:
- any internal data for the class will be managed on the heap instead of the stack and will provide some performance penalty, which could be negligible of course but not always
- there is nothing stopping the user to create other instances of these classes and that doesn't make sense as they are intended as libraries - just adds confusion especially to beginners
- methods like ```Application.Run``` cannot access the methods anymore
- there is nothing stopping you to write the full name of the method even if the methods sit within a standard module, if this is the coding style you prefer
- some people, me included nowadays, prefer not to add the module name because I prefer to keep lines of code shorter
- this adds extra maintenance because whenever you want to update to a newer version of the so-called libraries, you then need to go through the extra trouble to adjust code. For example an use of the ```AddressOf``` operator in a standard module would become a compile-time error when ported to a class. Another example would be the use of custom data types - I can see you've changed the visibility of ```REMOTE_MEMORY``` to ```Private``` in the ```MemoryLib``` class because you were forced by the compiler. However, that was on purpose intended to be ```Public``` because other projects like [WeakReference](https://github.com/cristianbuse/VBA-WeakReference) actually use it for 10%-15% performance gain like [here](https://github.com/cristianbuse/VBA-WeakReference/blob/9030d06f70a479cfc4041f45499f6254b2f35c0f/src/WeakReference.cls#L78).
- adds confusion to any new person looking at the code base as the implicit assumption is that a class is supposed to be used to create instances

Of course, I am not suggesting that you should change this. I am merely saying what considerations made me move away from this approach, just in case you haven't considered them all.

**Comment by Theadd** [comment]
Hi @cristianbuse ,

> My assumption is that you wanted to enforce the full name as in `JSON.Parse` hence avoiding to pollute the global namespace.

That's exactly the case of `FileSystemLib` and `ScreenLib` and a bunch of other "common" functions and utilities which I decided to move back to the project/app scope some time ago in order to focus ds3x on data collections only. So, when using intellisense, by typing `ds3x.`, it would display relevant stuff instead of all that garbage from before.
For the **JSON** case, used to work with js/ts/node, it was just personal. I even wish there were a simple `console.log` in VBA. This `JSON.cls` is quite altered from the original one, supporting ds3x's collections, opinionated pretty printing and allows bizarre stuff like `JSON.Parse("[, , , , 42]")     ' => Array(, , , , 42)`, which is not valid JSON. So, to anyone already using a JSON parser/serializer, it's better to stick with their own to expect the exact same output.
I'd prefer `JSON.cls` to remain a class module.

Regarding `MemoryLib.cls`, I agree. I'll move out my code in it to a separate class module or as private methods where needed and just add it as a git submodule, with some script to replace the one in `src/ds3x.accdb.src/modules/` at build time in the VCS. I might also do that with your `VBA-WeakReference` once I get to play with it.

Thank you!

**Comment by cristianbuse** [comment]
Hi @Theadd ,

> I'd prefer JSON.cls to remain a class module.

I remember testing the speed difference between a standard module JSON.bas and a class JSON.cls (identical code in both, with the exception of the PredeclaredID) and the difference was significant - if I remember corectly about 30-40% and it had something to do with the heap memory.

However, I've done this at least 5-6 years ago and I can't be sure about this. If I get a chance next week, i will try and test this again.

**Comment by cristianbuse** [comment]
Hi @Theadd ,

Just ran the speed test on JSON. The differences are negligible with the JSON converter found online. Just ignore my last comment.

Cheers!
